### PR TITLE
refactor: use PathProvider for user-data-dir and others

### LINKED ISSUE #29340: [Bug]: app.getPath('logs') points to the wrong directory
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Electron Version

13.0.1

### What operating system are you using?

Ubuntu

### Operating System Version

Ubuntu 20.04

### What arch are you using?

x64

### Last Known Working Electron version

_No response_

### Expected Behavior

When not setting any path manually `app.getPath('logs')` returns the actual path used for logging, which includes only one `<appname>`
* Windows - `C:\Users\<user>\AppData\Roaming\<appname>\logs`
* Linux - `/home/<user>/.config/<appname>/logs`

### Actual Behavior

When not setting any path manually `app.getPath('logs')` returns the path used for logging, but contains `<appname>` twice.
* Windows - `C:\Users\<user>\AppData\Roaming\<appname>\<appname>\logs`
* Linux - `/home/<user>/.config/<appname>/<appname>/logs`

### Testcase Gist URL

_No response_

### Additional Information

I tested this under Windows 10, as well as Ubuntu 20.04 and 21.04 using electron 10.1.3 and 13.0.1.

### PR DESCRIPTION
**Pull request by nornagon**
#### Description of Change
This extracts just the PathProvider refactor from #25089, for separate review.

Also fixes #29340.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes

#### Release Notes

Notes: Fixed `app.getPath('logs')` returning the wrong path on Linux and Windows.

### CODE REVIEW

**Comment by nornagon** [comment]
target/14-x-y because #25089 is too, and that depends on this.

**Comment by nornagon** [comment]
Possibly fixes #28701 as well?

File: shell/app/electron_main_delegate.cc [hunk 1]
```diff
@@ -107,28 +111,90 @@ void InvalidParameterHandler(const wchar_t*,
...
 #if defined(OS_MAC) || defined(OS_WIN)
-  cur = cur.Append(FILE_PATH_LITERAL("Crashpad"));
+      cur = cur.Append(FILE_PATH_LITERAL("Crashpad"));
+#else
+      cur = cur.Append(FILE_PATH_LITERAL("Crash Reports"));
+#endif
+      create_dir = true;
+      break;
+    case chrome::DIR_APP_DICTIONARIES:
+      base::debug::StackTrace().Print();
```
Comment (deepak1556) [inline shell/app/electron_main_delegate.cc:135]: Is this needed for prod ?

**Review by deepak1556 — APPROVED** [review]
LGTM with build failures addressed.

**Review by zcbenz — APPROVED** [review]
Looks good to me with above comments addressed.

**Pull request merged**